File: game.erl
%% Define suits and ranks as atoms
-define(SUITS, [club, diamond, spade, heart]).
-define(RANKS, [two, three, four, five, six, seven, eight, nine, ten, jack, queen, king, ace]).

%% Type Aliases (Informal, as Erlang doesn't have type aliases)
%% - Card: {Suit, Rank}
%% - Hand: List of Cards
%% - Deck: List of Cards

-record(player, {name :: string(), hand :: list()}).
-record(game, {deck :: list(), players :: list()}).

%% Function type equivalents (not strictly enforced in Erlang)
%% - Deal: fun(Deck) -> {Deck, Card}.
%% - PickupCard: fun(Hand, Card) -> Hand.

%% Example function: Deal a card from the deck
deal([Card | RestOfDeck]) -> {RestOfDeck, Card}.

%% Example function: Add a card to a hand
pickup_card(Hand, Card) -> [Card | Hand].

File: game.rs
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Suit {
    Club,
    Diamond,
    Spade,
    Heart,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Rank {
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Ten,
    Jack,
    Queen,
    King,
    Ace,
}

pub type Card = (Suit, Rank);
pub type Hand = Vec<Card>;
pub type Deck = Vec<Card>;

#[derive(Debug, Clone)]
pub struct Player {
    pub name: String,
    pub hand: Hand,
}

#[derive(Debug, Clone)]
pub struct Game {
    pub deck: Deck,
    pub players: Vec<Player>,
}

pub type Deal = fn(Deck) -> (Deck, Card);
pub type PickupCard = fn((Hand, Card)) -> Hand;
